src/MongoDB/ReadPreference.c

```diff
@@ -45,24 +45,43 @@ static PHP_METHOD(ReadPreference, __construct)
 
 	/* Separate the tagSets zval, since we may end up modifying it in
 	 * php_phongo_read_preference_prep_tagsets() below. */
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "l|a/!a!", &mode, &tagSets, &options) == FAILURE) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z|a/!a!", &mode, &tagSets, &options) == FAILURE) {
 		zend_restore_error_handling(&error_handling TSRMLS_CC);
 		return;
 	}
 	zend_restore_error_handling(&error_handling TSRMLS_CC);
 
-
-	switch(mode) {
-		case MONGOC_READ_PRIMARY:
-		case MONGOC_READ_SECONDARY:
-		case MONGOC_READ_PRIMARY_PREFERRED:
-		case MONGOC_READ_SECONDARY_PREFERRED:
-		case MONGOC_READ_NEAREST:
-			intern->read_preference = mongoc_read_prefs_new(mode);
-			break;
-		default:
-			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Invalid mode: %" PHONGO_LONG_FORMAT, mode);
+	if (Z_TYPE_P(mode) == IS_LONG) {
+		switch(Z_LVAL_P(mode)) {
+			case MONGOC_READ_PRIMARY:
+			case MONGOC_READ_SECONDARY:
+			case MONGOC_READ_PRIMARY_PREFERRED:
+			case MONGOC_READ_SECONDARY_PREFERRED:
+			case MONGOC_READ_NEAREST:
+				intern->read_preference = mongoc_read_prefs_new(Z_LVAL_P(mode));
+				break;
+			default:
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Invalid mode: %" PHONGO_LONG_FORMAT, Z_LVAL_P(mode));
+				return;
+		}
+	} else if (Z_TYPE_P(mode) == IS_STRING) {
+		if (strcmp(Z_STRVAL_P(mode), "primary") == 0) {
```

Comment:
I'll note that libmongoc also uses `strcasecmp()` for WriteConcern's "majority" string. Perhaps we want to do that as well in the WriteConcern constructor? We currently have:

```
if (strcmp(Z_STRVAL_P(w), PHONGO_WRITE_CONCERN_W_MAJORITY) == 0) {
```